### PR TITLE
[staking] ActiveCandidate Exclude Candidate with Expired Endorsement

### PR DESCRIPTION
**Pull request by envestcc**
# Description
Since the `SelfStake` field of a `Candidate` is not cleared when an endorsement expires, the current `ActiveCandidate` function does not remove expired candidates. This PR is aimed at fixing this issue.

## Type of change
Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/staking/protocol.go [hunk 1]
```diff
@@ -468,6 +468,55 @@ func (p *Protocol) Validate(ctx context.Context, act action.Action, sr protocol.
...
+		}
+		return false, err
+	}
+	// bucket is self-owned
+	if address.Equal(vb.Owner, cand.Owner) {
+		return true, nil
+	}
+	// bucket is not endorsed to the candidate
+	if !address.Equal(vb.Candidate, cand.Owner) {
+		return false, nil
```
Comment (CoderZhi) [inline action/protocol/staking/protocol.go:499]: if this happens, there must be something wrong with our implementation?

File: action/protocol/staking/protocol.go [hunk 2]
```diff
@@ -468,6 +468,55 @@ func (p *Protocol) Validate(ctx context.Context, act action.Action, sr protocol.
...
+		rp := rolldpos.MustGetProtocol(protocol.MustGetRegistry(ctx))
+		currentEpochNum := rp.GetEpochNum(srHeight)
+		if endorse.Status(rp.GetEpochLastBlockHeight(currentEpochNum+1)) == EndorseExpired {
+			return false, nil
+		}
+	case !errors.Is(err, state.ErrStateNotExist):
+		// other error
+		return false, err
+	default:
+		// endorsement does not exist
```
Comment (CoderZhi) [inline action/protocol/staking/protocol.go:515]: `return false, nil` or `return false, ErrNotExist`?

File: action/protocol/staking/protocol.go [hunk 3]
```diff
@@ -468,6 +468,55 @@ func (p *Protocol) Validate(ctx context.Context, act action.Action, sr protocol.
...
+		return false, nil
+	}
+	esr := NewEndorsementStateReader(csr.SR())
+	endorse, err := esr.Get(cand.SelfStakeBucketIdx)
+	switch {
+	case err == nil:
+		// endorsement exists and expired before end of next epoch
+		rp := rolldpos.MustGetProtocol(protocol.MustGetRegistry(ctx))
+		currentEpochNum := rp.GetEpochNum(srHeight)
+		if endorse.Status(rp.GetEpochLastBlockHeight(currentEpochNum+1)) == EndorseExpired {
```
Comment (CoderZhi) [inline action/protocol/staking/protocol.go:508]: can you elaborate it?

File: action/protocol/staking/protocol.go (hunk 1 above)
Comment (envestcc) [reply action/protocol/staking/protocol.go:499]: indeed, if clearing candidates is implemented after bucket endorsed to others

File: action/protocol/staking/protocol.go (hunk 3 above)
Comment (envestcc) [reply action/protocol/staking/protocol.go:508]: It is checking that only nodes which endorsement have not expired before the end of the next epoch will be considered as consensus nodes. This ensures that self-stake bucket is locked when make proposal, which facilitates the introduction of penalty mechanisms in the future.

File: action/protocol/staking/protocol.go (hunk 2 above)
Comment (envestcc) [reply action/protocol/staking/protocol.go:515]: it is unexpected behaviours, should return error

File: action/protocol/staking/protocol.go [hunk 4]
```diff
@@ -468,6 +468,52 @@ func (p *Protocol) Validate(ctx context.Context, act action.Action, sr protocol.
...
+		return true, nil
+	}
+	srHeight, err := csr.SR().Height()
+	if err != nil {
+		return false, errors.Wrap(err, "failed to get StateReader height")
+	}
+	vb, err := csr.getBucket(cand.SelfStakeBucketIdx)
+	if err != nil {
+		if errors.Is(err, state.ErrStateNotExist) {
+			return false, nil
```
Comment (CoderZhi) [inline action/protocol/staking/protocol.go:489]: SelfStakeBucketIndex shouldn't be an invalid one, so we may always return `false, err`

**Review by CoderZhi — APPROVED** [review]

File: action/protocol/staking/protocol.go [hunk 5]
```diff
@@ -492,7 +538,11 @@ func (p *Protocol) ActiveCandidates(ctx context.Context, sr protocol.StateReader
 			}
 			list[i].Votes.Add(list[i].Votes, contractVotes)
 		}
-		if list[i].SelfStake.Cmp(p.config.RegistrationConsts.MinSelfStake) >= 0 {
+		active, err := p.isActiveCandidate(ctx, c, list[i])
```
Comment (Liuhaai) [inline action/protocol/staking/protocol.go:541]: why are unactive cands included in L527?

Comment (envestcc) [reply action/protocol/staking/protocol.go:541]: There are two scenarios: 
- The self-stake bucket has been unstaked. 
- The endorsement has expired. 

Both situations belong to a registered delegate, but they cannot participate in consensus due to a lack of staking.

Comment (Liuhaai) [reply action/protocol/staking/protocol.go:519]: ok. It seems the state of candidates aren't tracked in the db(active/unactive). Pls add ur comments to explain why the list is filtered above L541

**Review by Liuhaai — APPROVED** [review]

File: action/protocol/staking/protocol.go [hunk 6]
```diff
@@ -472,6 +473,25 @@ func (p *Protocol) Validate(ctx context.Context, act action.Action, sr protocol.
...
 }
 
+func (p *Protocol) isActiveCandidate(ctx context.Context, csr CandidateStateReader, cand *Candidate) (bool, error) {
+	featureCtx := protocol.MustGetFeatureCtx(ctx)
+	// before delegate endorsement enabled, candidate is always active unless it's self-stake bucket is unstaked
+	if featureCtx.DisableDelegateEndorsement {
+		if cand.SelfStake.Cmp(p.config.RegistrationConsts.MinSelfStake) < 0 {
+			return false, nil
+		}
+		return true, nil
```
Comment (dustinxie) [inline action/protocol/staking/protocol.go:483]: just `return cand.SelfStake.Cmp(p.config.RegistrationConsts.MinSelfStake) >= 0, nil`

File: action/protocol/staking/protocol.go [hunk 7]
```diff
@@ -472,6 +473,25 @@ func (p *Protocol) Validate(ctx context.Context, act action.Action, sr protocol.
...
+	}
+
+	srHeight, err := csr.SR().Height()
+	if err != nil {
+		return false, errors.Wrap(err, "failed to get StateReader height")
+	}
+	rp := rolldpos.MustGetProtocol(protocol.MustGetRegistry(ctx))
+	currentEpochNum := rp.GetEpochNum(srHeight)
+	return csr.IsActiveCandidateAt(cand, rp.GetEpochLastBlockHeight(currentEpochNum+1))
```
Comment (dustinxie) [inline action/protocol/staking/protocol.go:492]: why should be `currentEpochNum+1`?

File: action/protocol/staking/protocol.go [hunk 8]
```diff
@@ -496,7 +516,11 @@ func (p *Protocol) ActiveCandidates(ctx context.Context, sr protocol.StateReader
 			}
 			list[i].Votes.Add(list[i].Votes, contractVotes)
 		}
-		if list[i].SelfStake.Cmp(p.config.RegistrationConsts.MinSelfStake) >= 0 {
+		active, err := p.isActiveCandidate(ctx, c, list[i])
+		if err != nil {
+			return nil, err
+		}
+		if active {
```
Comment (dustinxie) [inline action/protocol/staking/protocol.go:523]: 1. from the usage here, it is checking current height, not `currentEpochNum+1` height
2. to be safe, i think this check needs to be guarded by hard-fork flag

File: action/protocol/staking/candidate_statereader.go [hunk 9]
```diff
@@ -110,6 +112,41 @@ func (c *candSR) AllCandidates() CandidateList {
...
+	// self-owned bucket should not be unstaked
+	if address.Equal(vb.Owner, cand.Owner) {
+		return !vb.isUnstaked(), nil
+	}
+	// endorsed bucket should not be expired
+	if !address.Equal(vb.Candidate, cand.Owner) {
+		return false, nil
+	}
+	esr := NewEndorsementStateReader(c.SR())
+	endorse, err := esr.Get(cand.SelfStakeBucketIdx)
```
Comment (dustinxie) [inline action/protocol/staking/candidate_statereader.go:130]: this should be guarded by hard-fork flag? before 1.14, no such thing as endorsement exist

File: action/protocol/staking/protocol.go (hunk 7 above)
Comment (envestcc) [reply action/protocol/staking/protocol.go:492]: refer to https://github.com/iotexproject/iotex-core/pull/4062#discussion_r1461829192

File: action/protocol/staking/candidate_statereader.go (hunk 9 above)
Comment (envestcc) [reply action/protocol/staking/candidate_statereader.go:130]: hard-fork at https://github.com/iotexproject/iotex-core/pull/4062/files#diff-6bac7263ac5bdd5a72fb2bbb66bff76ebd0b794b0f998ed62bfafab7bba319e6R479

**Pull request merged**